Algorithm lookup

thecliguy opened the pull request:
New functionality added to perform a lookup of specified algorithm(s): ssh-audit.py --lookup=<alg>

jugmac00 commented:
Thank you very much for your pr!

A couple of thoughts:
This project recently started to apply both style and type linters. Both fail for your pr. Whereas the latter shows the errors, it is configured not to fail hard (ie. 0 return code although there are errors) (the reason is the `-` in front of the mypy call in the tox.ini).

You can run them locally with `tox` - https://tox.readthedocs.io/

@jtesta Maybe it would be a good idea to create a `CONTRIBUTING.md` file?

Generally, it would be a good idea to run `tox` once before creating a pr.

To reproduce the above mentioned linter failures, you could run

for the type checker
```
tox -e py38-mypy
```

for the style linter
```
tox -e py38-flake8
```

Also, you provided no tests. Whether tests for new code is wanted or not is on @jtesta to decide.

Also, I really like your example usage description in https://github.com/jtesta/ssh-audit/issues/52#issue-668202319

Unfortunately, once the issue is closed, this information is pretty much gone.

I really loved to see something like this in the documentation. That is another topic @jtesta should have a look and decide if and where documentation should be put. And it is always a good idea to add documentation along with the new feature, as we all know there will be never again the time to add documentation later on.

So, **tl/dr** - this is not about the pr in particular, but about general guidelines for new contributions.

Possible issues with this pr:
- flake8 fails
- mypy fails
- no tests
- no documentation

jugmac00 commented on ssh-audit.py:
```diff
@@ -3713,14 +3720,55 @@ def audit(aconf: AuditConf, sshv: Optional[int] = None, print_target: bool = Fal
+            try:
+                SSH2.KexDB.ALGORITHMS[alg_type][alg_name]
+                algorithms_dict.update( {alg_name : alg_type} ) 
+            except Exception:
```
Instead of a blank `Exception` you should use a more specific one, else this obfuscates other errors and makes debugging harder. Here you could use a `KeyError`. 

Catching a bare exception is also called the "most diabolical antipattern" in Python.
https://realpython.com/the-most-diabolical-python-antipattern/

thecliguy replied:
Thanks for the feedback, I'm going to try and rewrite that logic entirely so there's no reliance on an error handler.

I'm thinking of something like this:

```python
for k_algtype, v in SSH2.KexDB.ALGORITHMS.items():
    for k_algname, v1 in v.items():
        if k_algname in algorithms:
            algorithms_dict.update( {k_algname : k_algtype} ) 
```

jtesta commented on ssh-audit.py:
```diff
@@ -3713,14 +3720,55 @@ def audit(aconf: AuditConf, sshv: Optional[int] = None, print_target: bool = Fal
 
+def algorithm_lookup(alg_names):       
+    if not alg_names:    
+        out.fail("[exception] Missing argument: alg_names")
```
The program handles all final argument checking & validation in `AuditConf.from_cmdline()`.  It would be nice to move this checking there.

jtesta commented on ssh-audit.py:
```diff
@@ -3713,14 +3720,55 @@ def audit(aconf: AuditConf, sshv: Optional[int] = None, print_target: bool = Fal
+    if len(algorithms_not_found) > 0:
+        out.head('# unknown algorithms')
+        for algorithm_not_found in algorithms_not_found:
+            out.fail(algorithm_not_found) 
```
I suppose `retval = PROGRAM_RETVAL_FAILURE` would be appropriate when the algorithm isn't found.

thecliguy replied:
@jtesta Would you like `PROGRAM_RETVAL_FAILURE` to be returned if any one of the specified algorithms could not be found? Or only if **all** of the specified algorithms could not be found?

thecliguy replied:
How about if all the specified algorithms could not be found then return `PROGRAM_RETVAL_FAILURE`, if some (but not all) of the specified algorithms could not be found then return `PROGRAM_RETVAL_WARNING`, else return `PROGRAM_RETVAL_GOOD` ?

jtesta replied:
The way it works now, if at least one algorithm fails, the program returns failure overall (it doesn't matter if all other algs are good).

So perhaps we do the same thing here: if at least one is unknown (= failure), the result should be failure overall.

jtesta commented on ssh-audit.py:
```diff
@@ -3713,14 +3720,55 @@ def audit(aconf: AuditConf, sshv: Optional[int] = None, print_target: bool = Fal
 
+    if aconf.lookup != '':
+        algorithm_lookup(aconf.lookup)
+        sys.exit(0)
```
Now we can do:
```
retval = algorithm_lookup(aconf.lookup)
sys.exit(retval)
```

jtesta commented on ssh-audit.py:
```diff
@@ -3713,14 +3720,55 @@ def audit(aconf: AuditConf, sshv: Optional[int] = None, print_target: bool = Fal
+        
+        if len(algorithms_dict_filtered.items()) > 0:
+            title = alg_types.get(alg_type)
+            output_algorithms(title, adb, alg_type, algorithms_dict_filtered.keys(), unknown_algorithms, False, PROGRAM_RETVAL_GOOD, padding)
```
The program should return the appropriate exit code when it ends.  So we should track the return value from `output_algorithms()`.

This can be done by initializing `retval = PROGRAM_RETVAL_GOOD`, then continually updating it with `retval = output_algorithms([...], retval, [...])`.  Then return `retval` to the caller.

jtesta commented:
> flake8 fails
> mypy fails
> no tests
> no documentation

I consider passing code tests, new unit tests, and documentation to be great to have, but not strictly required.  The more hard requirements we put in place, the less people are encouraged to submit PRs, in my opinion.  I'd rather be more encouraging to newcomers and get more involvement. Then later, repeat contributors can take on more responsibility.

So there you have it, @cliguy.  If you've got the time and motivation to fix the tox tests, perhaps write a unit test, update the usage in README.md, and update the [man page](https://github.com/jtesta/ssh-audit/blob/dev/ssh-audit.1), please do!  Otherwise, I'd be happy to accept your work and finish those parts up myself.

thecliguy commented:
@jtesta @jugmac00 I've updated the PR...

In `algorithm_lookup`:
  * I've replaced some nested loops with list and dictionary comprehension.
  * The code has been revised so there's no dependence of an error handler as part of the logic.

I've updated the `Usage` section in `README.md`. I was going to go further and add a example along with a screenshot but I noticed that the existing screenshots are not located within the project. Perhaps it would be a good idea to add a directory to the folder for assets like screenshots? I've prepared a screenshot if you want to use it, see attached:

![ssh-audit-lookup-example](https://user-images.githubusercontent.com/23440389/89103053-cea47c80-d406-11ea-8eaa-b05ca366f491.png)

This time, before pushing I ran `~/.local/bin/tox -e py38-mypy` and `~/.local/bin/tox -e py38-flake8` against my local copy of the code.

I haven't written tests because it's beyond my knowledge. If you do write tests for this code, I will of course study them so I can learn and hopefully contribute PRs with tests in future.

Please can you review the code and let me know if any further changes are needed.
